projects/data-grid/src/lib/directives/column-resize.directive.ts

```diff
@@ -0,0 +1,99 @@
+import { AfterViewInit, Directive, ElementRef, EventEmitter, Input, Output, Renderer2 } from '@angular/core';
+import { DataGridColumnWithId } from '../models';
+import * as R from 'ramda';
+import { EventTypes } from '../models/event-types';
+import { EventTargetTypes } from '../models/event-target-types';
+
+const CHILDREN_PADDINGS_WIDTH = 20;
+
+@Directive({
+  selector: '[ngrxColumnResize]'
+})
+export class ColumnResizeDirective implements AfterViewInit {
+  @Input() column: DataGridColumnWithId;
+
+  @Output() readonly columnResizing = new EventEmitter<DataGridColumnWithId>();
+  @Output() readonly columnResized = new EventEmitter<DataGridColumnWithId>();
+
+  columnInResizeMode = false;
+  resizeStartPositionX: number;
+  minColumnWidth: number;
+
+  constructor(
+    private elementRef: ElementRef,
+    private renderer: Renderer2,
+  ) {
+  }
+
+  ngAfterViewInit(): void {
+    this.minColumnWidth = this.calculateMinColumnWidth();
+  }
+
+  get element() {
+    return this.elementRef.nativeElement;
+  }
+
+  get elementWidth() {
+    return this.element.getBoundingClientRect().width;
+  }
+
+  onMouseDown(event: MouseEvent) {
+    event.stopPropagation();
+    this.columnInResizeMode = true;
+    this.resizeStartPositionX = event.x;
+    this.column = this.updateColumnWidth();
+    this.registerMouseEvents();
+  }
+
+  onMouseUp(event: MouseEvent) {
+    event.stopPropagation();
+    if (this.columnInResizeMode) {
+      this.columnResized.emit(this.resizeColumn(event.x));
+      this.columnInResizeMode = false;
+    }
+  }
+
+  onMouseMove(event: MouseEvent) {
+    if (this.columnInResizeMode) {
+      this.columnResizing.emit(this.resizeColumn(event.x));
+    }
+  }
+
+  private registerMouseEvents() {
+    const unregisterMouseMove = this.renderer.listen(EventTargetTypes.Document, EventTypes.MouseMove, (mouseMoveEvent: MouseEvent) => {
+      this.onMouseMove(mouseMoveEvent);
+    });
+
+    const unregisterContextMenu = this.renderer.listen(EventTargetTypes.Document, EventTypes.ContextMenu, (contextmenu: MouseEvent) => {
+      contextmenu.preventDefault();
+    });
+
+    const unregisterMouseUp = this.renderer.listen(EventTargetTypes.Document, EventTypes.MouseUp, (mouseUpEvent: MouseEvent) => {
+      this.onMouseUp(mouseUpEvent);
+      unregisterMouseUp();
+      unregisterMouseMove();
+      unregisterContextMenu();
+    });
+  }
+
+  private updateColumnWidth() {
+    return R.mergeRight(this.column, {width: this.elementWidth});
+  }
+
+  private resizeColumn(currentPositionX: number): DataGridColumnWithId {
+    return R.mergeRight(this.column, {width: this.calculateColumnWidth(currentPositionX)});
```

Comment:
Why are we changing the width of the column here? Shouldn't we just calculate the new width and update the width of the column in reducer.